CNV-10445 [1926199] VM Migration of mixed CPU (AMD,Intel) (#30340)
openshift/openshift-docs: virt/install/preparing-cluster-for-virt.adoc

* To use proxy with {VirtProductName}, you must xref:../../operators/admin/olm-configuring-proxy-support.adoc#olm-configuring-proxy-support[configure proxy support in Operator Lifecycle Manager].

> * When migrating virtual machines to a cluster, ensure that all nodes in a single pool use the same type of CPU. For example, nodes in one pool can use AMD processors and nodes in another pool can use Intel processors, and so on. Avoid mixing different CPU types in the same pool. 

Reviewer: I think the tense is confusing ("When migrating..." vs. "Before you can migrate"). I also think this needs to be a bit more direct. 

(Also, I searched the repo to double check and we generally use "to" in cases like this.)

Proposed rewrite:

```suggestion
* Before you can migrate virtual machines to a cluster, you must ensure that all nodes in a single pool use the same type of CPU. Do not mix different CPU types in the same pool. For example, place nodes that use AMD processors in one pool and place nodes that use Intel processors in a different pool.
```

Reviewer 2: The "before" method sounds good to me because this section is part of the prerequisites so it fits.
This issue is right about 2.4 -2.5 (I think Its solved on 2.6 but I will check)

Author: We can go with Before if that makes sense. The other prereqs in the assembly do not "before" as they are written more as general guidelines rather than sequential steps in a task. 

I do not think "you must" is needed as "ensure" is a directive verb. I also think the example should come immediately after the initial guideline. The other reviewers I consulted (Catherine and Nunzy) seem to feel the same way on these two points.

Rewriting as:
_Before you can migrate virtual machines to a cluster, ensure that all nodes in a single pool use the same type of CPU. For example, place nodes that use AMD processors in one pool and place nodes that use Intel processors in a different pool. Do not mix different CPU types in the same pool._

Author: Also dropping "can" --- it is not needed.

Reviewer 2: Sounds good to me :)

Reviewer 2: One thing , maybe we should write INTEL(all capital)  , we did write AMD , do we want to stay on the same "line" ?

Author: Hi @kobi86, AMD is capitalized as it is the commonly used acronym to describe Advanced Micro Devices, Intel is the  company name, that's the difference. No worries. :-) 

Should this be published for 2.6 or only for 2.4 and 2.5?

 Bob

Reviewer 2: 2.4 and 2.5 for sure 

about 2.6 - I would like to check to make sure

> * Before you migrate virtual machines to a cluster, ensure that all nodes in a single pool use the same type of CPU. For example, place nodes that use AMD processors in one pool and place nodes that use Intel processors in a different pool. Do not mix different CPU types in the same pool.

Reviewer 3: /hold

I find this text confusing. What is to "migrate virtual machines to a cluster"? You can either import a VM to cluster, or live-migrate a VM *in* a cluster. I suppose this release note refers to the latter, but typical users do not choose to migrate, they can do nothing "before" migrations. Also, I don't know what is the "pool" this text refers to.

Author: OK, thanks @dankenigsberg.  I can rewrite this with the new information supplied by Israel re: labelselectors and affinity.

Reviewer 3: Great. How can we make sure to use only well-defined  concepts (unlike "cpu pools")?

Author: A few things I can think of:

Reviewer 2: Thanks for all the help :)

Author: No problem @kobi86. It helps me to learn about these concepts (labelselectors, etc) so there is good in that :-)

Reviewer 3: @kobi86 @bgaydosrh can you drop this "cpu pool" concept from the BZ and PR?

* To use proxy with {VirtProductName}, you must xref:../../operators/admin/olm-configuring-proxy-support.adoc#olm-configuring-proxy-support[configure proxy support in Operator Lifecycle Manager].

> * When importing a virtual machines to a cluster or live migrating a virtual machine within a cluster, label nodes based on the CPU vendor so that all virtual machines in a pod use the same type of CPU. For example, virtual machines using AMD processors should not be placed on the same node as virtual machines that use Intel processors. xref:../../nodes/scheduling/nodes-scheduler-node-affinity.adoc#nodes-scheduler-node-affinity-configuring-required_nodes-scheduler-node-affinity[Configure a required node affinity rule] using the `Vendor` label to ensure virtual machines with different CPUs are not mixed on the same node.

Reviewer 3: There's no "when live migrating". The person defining the VM (be it via import, wizard or yaml) does not control or even know when their VM is going to migrate. Also, there is no "all virtual machines in a pod". The text should
 * state the problem: two vendors in a cluster is a mess
 * request to label each node based on its vendor
 * request to label each VM, effectively slicing the cluster in two

suggestion:
If your cluster consists more than a single CPU vendor, VMs may attempt to live-migrate from one to another and likely to fail. To avoid this, you should label your nodes with a `Vendor=Intel` or `Vendor=AMD` label (add link, please) and set node affinity on your VMs so that each can run only on one vendor.

Author: OK, thanks @dankenigsberg. By way of explanation (not excuse) I got the first sentence from your prior comment _You can either import a VM to cluster, or live-migrate a VM in a cluster. I suppose this release note refers to the latter, but typical users do not choose to migrate, they can do nothing "before" migrations._ I also made the assumption that since "before" was not correct as I'd suspected, "when" was appropriate. My bad. Revising as you have suggested. Here is my wording based on what you just gave me. If you approve, please remove the hold label. Thanks again.

_* If your cluster uses virtual machines from multiple CPU vendors, live migration failures can occur. For example, a virtual machine with an AMD CPU may attempt to live-migrate to a virtual machine with an Intel CPU and likely fail migration. To avoid this, label nodes with a vendor-specific label, such as `Vendor=Intel` or `Vendor=AMD`, and set node affinity on your virtual machines to ensure successful migration. See xref:../../nodes/scheduling/nodes-scheduler-node-affinity.adoc#nodes-scheduler-node-affinity-configuring-required_nodes-scheduler-node-affinity[Configuring a required node affinity rule] for more information._

It sounds like we have supported the `Vendor=` label in OCP for some time. If so, I will target this for 2.4, 2.5, 2.6 and 4.8 if that makes sense (I will need to check to see if we are still updating 2.4). 

cc: @ILpinto @ctomasko @nunzy1

> * If your cluster uses virtual machines from multiple CPU vendors, live migration failures can occur. For example, a virtual machine with an AMD CPU may attempt to live-migrate to a virtual machine with an Intel CPU and likely fail migration. To avoid this, label nodes with a vendor-specific label, such as `Vendor=Intel` or `Vendor=AMD`, and set node affinity on your virtual machines to ensure successful migration. See xref:../../nodes/scheduling/nodes-scheduler-node-affinity.adoc#nodes-scheduler-node-affinity-configuring-required_nodes-scheduler-node-affinity[Configuring a required node affinity rule] for more information.

Reviewer 3: ```suggestion
* If your cluster uses worker nodes from multiple CPU vendors, live migration failures can occur. For example, a virtual machine with an AMD CPU may attempt to live-migrate to a node with an Intel CPU and likely fail migration. To avoid this, label nodes with a vendor-specific label, such as `Vendor=Intel` or `Vendor=AMD`, and set node affinity on your virtual machines to ensure successful migration. See xref:../../nodes/scheduling/nodes-scheduler-node-affinity.adoc#nodes-scheduler-node-affinity-configuring-required_nodes-scheduler-node-affinity[Configuring a required node affinity rule] for more information.
```

Reviewer 3: /hold cancel

* To use proxy with {VirtProductName}, you must xref:../../operators/admin/olm-configuring-proxy-support.adoc#olm-configuring-proxy-support[configure proxy support in Operator Lifecycle Manager].

> * If your cluster uses worker nodes from multiple CPU vendors, live migration failures can occur. For example, a virtual machine with an AMD CPU may attempt to live-migrate to a node with an Intel CPU and likely fail migration. To avoid this, label nodes with a vendor-specific label, such as `Vendor=Intel` or `Vendor=AMD`, and set node affinity on your virtual machines to ensure successful migration. See xref:../../nodes/scheduling/nodes-scheduler-node-affinity.adoc#nodes-scheduler-node-affinity-configuring-required_nodes-scheduler-node-affinity[Configuring a required node affinity rule] for more information.

Reviewer 4: s/may/might

Author: Done.

Author: Done
